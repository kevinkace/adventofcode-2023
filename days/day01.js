const numMap = {
    one   : "o1e",
    two   : "t2o",
    three : "t3e",
    four  : "f4r",
    five  : "f5e",
    six   : "s6x",
    seven : "s7n",
    eight : "e8t",
    nine  : "n9e",
};

export function parse(str) {
    return str.split("\n");
}

/**
 * Get first and last digit in string, and concat to a 2 digit number
 * @param {string} row - eg `"abc123"`
 * @returns {number} - eg `13`
 */
export function getNum(row) {
    const nums = row.match(/\d/g);

    return nums ?
        parseInt(`${nums.at(0)}${nums.at(-1)}`, 10) :
        0;
}

/**
 * Sum all numbers in input
 * @param {string} input - eg `"abc123\nabc456"`
 * @returns {number} - eg `59`, because `13 + 46 = 59`
 */
export function addRowToSum(sum, row) {
    const num =  getNum(row);

    return sum + num;
}

export function replaceStringsWithNumbers(str) {
    Object.entries(numMap).forEach(([ word, num ]) => {
        str = str.replaceAll(word, num);
    });

    return str;
}
