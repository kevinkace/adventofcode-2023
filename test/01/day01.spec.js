import { readFile }      from "node:fs/promises";
import { fileURLToPath } from "node:url";
import path              from "node:path";

import { expect, test, describe } from "vitest";

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

const input    = await readFile(path.join(__dirname, "input.txt"), "utf8");
const sample01 = await readFile(path.join(__dirname, "sample01.txt"), "utf8");
const sample02 = await readFile(path.join(__dirname, "sample02.txt"), "utf8");

import { getNum, replaceStringsWithNumbers, addRowToSum, parse, addRowToSum } from "../../days/day01.js";

describe("day 01", () => {
    test("gets nums from row", () => {
        expect(getNum("abc123")).toEqual(13);
        expect(getNum("abc1")).toEqual(11);
        expect(getNum("abc")).toEqual(0);
        expect(getNum("threeight")).toEqual(0);
    });

    test("can sum rows", () => {
        expect(addRowToSum(0, "abc123")).toEqual(13);
        expect(addRowToSum(13, "abc456")).toEqual(59);
    });

    test("answer part 1", () => {
        let result = parse(sample01)
            .reduce(addRowToSum, 0);

        expect(result).toEqual(142);

        result = parse(input)
            .reduce(addRowToSum, 0);

        expect(result).toMatchSnapshot();
    });

    test("answer part 2", async () => {
        let result = parse(sample02)
            .map(replaceStringsWithNumbers)
            .reduce(addRowToSum, 0);

        expect(result).toEqual(281);

        result = parse(input)
            .map(replaceStringsWithNumbers)
            .reduce(addRowToSum, 0);

        expect(result).toMatchSnapshot();
    });
});
