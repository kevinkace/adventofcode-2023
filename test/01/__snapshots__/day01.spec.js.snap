// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 01 > answer part 1 1`] = `54877`;

exports[`day 01 > answer part 2 1`] = `54100`;
